File: app/src/main/java/projects/kyle/ledcontroller/MainContent/MainActivityAdapter.java
package projects.kyle.ledcontroller.MainContent;

import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import java.util.ArrayList;

import projects.kyle.ledcontroller.R;

public class MainActivityAdapter extends RecyclerView.Adapter<MainActivityAdapter.MainActivityViewHolder> {

    private ArrayList<MainListItem> viewList;

    public MainActivityAdapter(ArrayList<MainListItem> viewList){
        this.viewList = viewList;
    }



    @NonNull
    @Override
    public MainActivityViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {

        View v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.main_list_item, viewGroup, false);
        v.setOnClickListener(viewList.get(i).listener);

        MainActivityViewHolder vh = new MainActivityViewHolder(v, viewGroup);
        View prev = LayoutInflater.from(viewGroup.getContext()).inflate(viewList.get(i).preview, viewGroup, false);
        viewList.get(i).callback.callback(prev);
        vh.preview.addView(prev);

        return vh;
    }

    @Override
    public int getItemViewType(int position){
        return position;
    }
    

    @Override
    public void onBindViewHolder(@NonNull MainActivityViewHolder mainActivityViewHolder, int i) {
        mainActivityViewHolder.menuLabel.setText(viewList.get(i).label);

    }

    @Override
    public int getItemCount() {
        return viewList.size();
    }

    public static class MainActivityViewHolder extends RecyclerView.ViewHolder {

        public TextView menuLabel;
        public FrameLayout preview;
        public ViewGroup viewGroup;

        public MainActivityViewHolder(View view, ViewGroup viewGroup) { super(view);
           this.viewGroup = viewGroup;
           menuLabel = (TextView) view.findViewById(R.id.main_list_item_label);
           preview = (FrameLayout) view.findViewById(R.id.main_list_item_preview);
        }
    }
}

File: app/src/main/java/projects/kyle/ledcontroller/MainContent/MainActivityAdapter.kt
package projects.kyle.ledcontroller.MainContent

import android.support.constraint.ConstraintLayout
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.TextView

import java.util.ArrayList

import projects.kyle.ledcontroller.R

class MainActivityAdapter(private val viewList: ArrayList<MainListItem>) : RecyclerView.Adapter<MainActivityAdapter.MainActivityViewHolder>() {
    override fun getItemCount(): Int {
        return viewList.size
    }

    override fun  onCreateViewHolder(viewGroup: ViewGroup, i: Int): MainActivityViewHolder {

        val v = LayoutInflater.from(viewGroup.context).inflate(R.layout.main_list_item, viewGroup, false)
        v.setOnClickListener(viewList[i].listener)

        val vh = MainActivityViewHolder(v, viewGroup)
        val prev = LayoutInflater.from(viewGroup.context).inflate(viewList[i].preview, viewGroup, false)
        viewList[i].callback?.callback(prev)
        vh.preview.addView(prev)

        return vh
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }


    override fun onBindViewHolder(mainActivityViewHolder: MainActivityViewHolder, i: Int) {
        mainActivityViewHolder.menuLabel.text = viewList[i].label

    }

    class MainActivityViewHolder(view: View, var viewGroup: ViewGroup) : RecyclerView.ViewHolder(view) {

        var menuLabel: TextView
        var preview: FrameLayout

        init {
            menuLabel = view.findViewById<View>(R.id.main_list_item_label) as TextView
            preview = view.findViewById<View>(R.id.main_list_item_preview) as FrameLayout
        }
    }
}

